src/physics/em/detail/KleinNishina.cu

```diff
@@ -0,0 +1,81 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file KleinNishina.cu
+//---------------------------------------------------------------------------//
+#include "KleinNishina.hh"
+
+#include "base/KernelParamCalculator.cuda.hh"
+#include "physics/base/SecondaryAllocatorView.hh"
+#include "physics/base/ModelInterface.hh"
+#include "physics/base/ParticleTrackView.hh"
+#include "physics/base/PhysicsTrackView.hh"
+#include "random/cuda/RngEngine.hh"
+#include "KleinNishinaInteractor.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+// KERNELS
+//---------------------------------------------------------------------------//
+/*!
+ * Interact using the Klein-Nishina model on applicable tracks.
+ */
+__global__ void klein_nishina_interact_kernel(const KleinNishinaPointers  kn,
+                                              const ModelInteractPointers ptrs)
+{
+    auto tid = celeritas::KernelParamCalculator::thread_id();
+    if (tid.get() >= ptrs.states.size())
+        return;
+
+    SecondaryAllocatorView allocate_secondaries(ptrs.secondaries);
+    ParticleTrackView particle(ptrs.params.particle, ptrs.states.particle, tid);
+    PhysicsTrackView  physics(ptrs.params.physics,
+                             ptrs.states.physics,
+                             particle.def_id(),
+                             MaterialDefId{},
+                             tid);
+    RngEngine         rng(ptrs.states.rng, tid);
```

Comment:
Just to be sure: would calling `RngEngine` before we know if the interaction is going to happen result in non-reproducibility?

Reply:
This isn't calling it; it's instantiating it, which shouldn't do any actual work. I thought I changed the code to put it after the check on the model -- I'll restore that change, thanks for the check.